### PR TITLE
[Resolve #1114,#886,#491] Project Dependencies part 2: Resolvable role_arn and template_bucket_name

### PR DESCRIPTION
**Pull request by jfalkenstein**
This is the second in a series of pull requests that addresses #1114 , allowing Sceptre to manage its own dependencies.

## In this PR:
* The ResolvableValueProperty is created, creating a property that can resolve to a single value (not a list or dict)
* role_arn and template_bucket_name are now fully resolvable properties
    - role_arn also resolves the issue #886 
    - template_bucket_name also provides a solution to the issue #491 
* If template_bucket_name is set to None on a stack, it will be interpreted as if it wasn't there at all. _This will be important in future Pull Requests so that if template_bucket_name is inherited from the stack_group, it won't actually create a circular dependency on the stack that outputs that bucket name._
* Documentation is clarified on setting stack dependencies on a StackGroup to be shared across all stacks in that group as well as a warning about circular dependencies.

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [x] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

**Comment by jfalkenstein** [comment]
@schlichtanders, This addresses what you requested in #886.

@craighurley This addresses what you requested in #491 (well, it makes it possible to setup whatever kind of bucket you want).

File: sceptre/stack.py [hunk 1]
```diff
@@ -116,7 +116,12 @@ class Stack(object):
     parameters = ResolvableContainerProperty("parameters")
     sceptre_user_data = ResolvableContainerProperty("sceptre_user_data")
     notifications = ResolvableContainerProperty("notifications")
-    tags = ResolvableContainerProperty('tags')
+    tags = ResolvableContainerProperty("tags")
+
+    s3_details = ResolvableContainerProperty("s3_details")
```
Comment (zaro0508) [inline sceptre/stack.py:121]: why is this a resolvable property?  it's not a key in stack configs is it?

Comment (jfalkenstein) [reply sceptre/stack.py:121]: So the template_bucket_name is accessed off of this property in the template class. So, in order to make template_bucket_name resolvable in all cases, the s3_details needed to be resolvable too.

File: sceptre/config/reader.py [hunk 2]
```diff
@@ -475,7 +475,9 @@ def _collect_s3_details(stack_name, config):
         :rtype: dict
         """
         s3_details = None
-        if "template_bucket_name" in config:
+        # If the config explicitly sets the template_bucket_name to None, we don't want to enter
+        # this conditional block.
+        if config.get("template_bucket_name") is not None:
```
Comment (zxiiro) [inline sceptre/config/reader.py:480]: I think we can shorten this to `if config.get("template_bucket_name"):`?

I believe config.get() defaults to `None` if it's not found and if template_bucket_name is explicity None then the condition wouldn't enter the block.

File: sceptre/template.py [hunk 3]
```diff
@@ -215,7 +215,8 @@ def get_boto_call_parameter(self):
         :returns: The boto call parameter for the template.
         :rtype: dict
         """
-        if self.s3_details:
+        # If bucket_name is set to None, it should be ignored and not uploaded.
+        if self.s3_details and self.s3_details.get("bucket_name") is not None:
```
Comment (zxiiro) [inline sceptre/template.py:219]: I believe the `is not None` can be shortend to simply `if self.s3_details and self.s3_details.get("bucket_name")` here?

**Review by zxiiro — COMMENTED** [review]
Really minor comment really about the `is not None` in if conditions. Either way I think this looks fine to me.

File: sceptre/config/reader.py (hunk 2 above)
Comment (jfalkenstein) [reply sceptre/config/reader.py:480]: You're right that None is falsey. However, I like to try to be as explicit in these checks as possible, since other values could be falsey too, like empty strings. I suppose in this case, we might want to treat an empty string the same as `None`. Nonetheless, when one is not explicit about what they're checking for, it can lead future maintainers to wonder about what sorts of values could be expected here.

**Comment by jfalkenstein** [comment]
Alright @zaro0508 , @zxiiro this PR is now fully ready for review. You both had made some comments on this PR previously (and I've responded to those). I've added in proper docs and integration tests on this PR, so it should now be ready for review.

**This PR can be merged and deployed to master without requiring any changes from parts 3 and onward**. However, this PR _is_ pointed at the Part 1 branch because it builds on the changes in that one.

File: docs/_source/docs/stack_group_config.rst [hunk 4]
```diff
@@ -66,6 +66,13 @@ supplies the template to Boto3 via the ``TemplateBody`` argument. Templates
 supplied in this way have a lower maximum length, so using the
 ``template_bucket_name`` parameter is recommended.
 
+.. warning::
+
+   If you resolve ``template_bucket_name`` using the ``!stack_output``
+   resolver on a StackGroup, the stack that outputs that bucket name *cannot* be
+   defined in that StackGroup. Otherwise, a circular dependency will exist and Sceptre
+   will raise an error when attempting any Stack action.
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply docs/_source/docs/stack_group_config.rst:74]: Yeah, I can add another line indicating the correct way to do it.

**Comment by jfalkenstein** [comment]
@zaro0508 , I gave two examples of project structure supporting referencing project dependencies with `!stack_output`.

**Review by zaro0508 — APPROVED** [review]
